tests/integration/test_tracking.py

```diff
@@ -1062,6 +1064,311 @@ def allocating_function(allocator, amount, stop_flag):
+        all_allocations = list(reader.get_allocation_records())
+        assert len(all_allocations) >= 1
+        assert not list(
+            filter_relevant_allocations(reader.get_temporary_allocation_records())
+        )
+
+    def test_thread_allocations_multiple_threads(self, tmpdir):
+
+        # GIVEN
+        def allocating_function(allocator, amount, stop_flag):
+            allocator.posix_memalign(amount)
+            allocator.free()
+            allocator.posix_memalign(amount)
+            allocator.free()
+            # We need a barrier as pthread might reuse the same thread ID if the first thread
+            # finishes before the other one starts
+            stop_flag.wait()
+
+        # WHEN
+        alloc1 = MemoryAllocator()
+        stop_flag1 = threading.Event()
+        alloc2 = MemoryAllocator()
+        stop_flag2 = threading.Event()
+        output = Path(tmpdir) / "test.bin"
+        with Tracker(output):
+            t1 = threading.Thread(
+                target=allocating_function, args=(alloc1, 2048, stop_flag1)
+            )
+            t1.start()
+            t2 = threading.Thread(
+                target=allocating_function, args=(alloc2, 2048, stop_flag2)
+            )
+            t2.start()
+
+            stop_flag1.set()
+            t1.join()
+            stop_flag2.set()
+            t2.join()
+
+        # THEN
+        reader = FileReader(output)
+        all_allocations = [
+            record
+            for record in reader.get_allocation_records()
+            if record.allocator == AllocatorType.POSIX_MEMALIGN
+        ]
+        assert len(all_allocations) == 4
+
+        temporary_allocation_records = (
+            record
+            for record in reader.get_temporary_allocation_records(merge_threads=False)
+            if record.allocator == AllocatorType.POSIX_MEMALIGN
+        )
+        # Group the allocations per thread
+        records = collections.defaultdict(list)
+        for record in temporary_allocation_records:
+            records[record.tid].append(record)
+        assert len(records.keys()) == 2
+        # Each thread should have 4096 bytes total allocations
+        for tid, allocations in records.items():
+            assert sum(allocation.size for allocation in allocations) == 4096
+
+    def test_intertwined_temporary_allocations_in_threads(self, tmpdir):
+        """This test checks that temporary allocations are correctly detected
+        when they happen in different threads with interleaved calls to malloc
+        and free.
+
+        Note that there is not an easy way to synchronize this test because
+        using condition variables may allocate, disrupting the test as there
+        will be other spurious allocations between our call to posix_memalign
+        and the call to free.
+
+        The best effort is that both threads do a bunch of calls to
+        posix_memalign and free in a loop and hopefully some of them will be
+        interleaved.
```

Comment:
How about we use `time.sleep()` instead? At the cost of making the test take 750ms to run, we could make
```
Thread 1    Thread 2
------ time=0ms -------
allocate    sleep(0.25)
sleep(0.5)
------ time=250ms -----
            allocate
            sleep(0.5)
------ time=500ms -----
    free
------ time=750ms -----
            free
```

Sure, that's racy - but so is this approach, so I'm not sure it's any worse.

I bet we could enforce that order by spinlocking on a Python value, actually... What if we did:

```py
thread1_allocated = False
thread1_should_free = False

def thread1_body(allocator):
    global thread_1_allocated
    allocator.posix_memalign(amount)
    thread_1_allocated = True
    while not thread_1_should_free:
        pass
    allocator.free()

def thread2_body(allocator):
    global thread1_should_free
    while not thread1_allocated:
        pass
    allocator.alloc()
    allocator.free()
    thread1_should_free = True
```

I think that's not racy (thanks to the GIL) and I think no unexpected allocations can happen in there.

Reply:
Ok, let's go with the second version spinlocking on a python value